app/controllers/case_contacts/form_controller.rb

```diff
@@ -0,0 +1,145 @@
+  include Wicked::Wizard
+
+  before_action :set_progress
+  before_action :require_organization!
+  after_action :verify_authorized
+
+  steps(*CaseContact::FORM_STEPS)
+
+  # wizard_path
+  def show
+    @case_contact = CaseContact.find(params[:case_contact_id])
+    authorize @case_contact
+    get_cases_and_contact_types
+    @page = wizard_steps.index(step) + 1
+    @total_pages = steps.count
+
+    render_wizard
+    wizard_path
+  end
+
+  def update
+    @case_contact = CaseContact.find(params[:case_contact_id])
+    authorize @case_contact
+    params[:case_contact][:status] = step.to_s unless @case_contact.active?
+    remove_unwanted_contact_types
+    remove_nil_draft_ids
+    if @case_contact.update(case_contact_params)
+      respond_to do |format|
+        format.html {
+          if step == steps.last
+            finish_editing
+          else
+            render_wizard @case_contact, {}, {case_contact_id: @case_contact.id}
+          end
+        }
+        format.json { head :ok }
+      end
+    else
+      respond_to do |format|
+        format.html {
+          get_cases_and_contact_types
+          render step
+        }
+        format.json { head :internal_server_error }
+      end
+    end
+  end
+
+  private
+
+  def get_cases_and_contact_types
+    @casa_cases = policy_scope(current_organization.casa_cases)
+    @casa_cases = @casa_cases.where(id: @case_contact.casa_case_id) if @case_contact.active?
+
+    @selected_case_contact_types = @casa_cases.flat_map(&:contact_types)
+
+    @current_organization_groups =
+      if @selected_case_contact_types.present?
+        @selected_case_contact_types.map(&:contact_type_group).uniq
+      else
+        current_organization.contact_types_by_group
+      end
+  end
+
+  def finish_editing
+    message = ""
+    send_reimbursement_email(@case_contact)
+    if @case_contact.active?
+      message = @case_contact.decorate.form_updated_message
+    else
+      message = "Case #{"contact".pluralize(@case_contact.draft_case_ids.count)} successfully created."
+      create_additional_case_contacts(@case_contact)
+      first_casa_case_id = @case_contact.draft_case_ids.slice(0)
+      @case_contact.update(status: "active", draft_case_ids: [first_casa_case_id], casa_case_id: first_casa_case_id)
+    end
+    update_volunteer_address(@case_contact)
+    flash[:notice] = message
+    redirect_back_to_referer(fallback_location: case_contacts_path(success: true))
+  end
+
+  def send_reimbursement_email(case_contact)
+    if case_contact.should_send_reimbursement_email?
+      SupervisorMailer.reimbursement_request_email(case_contact.creator, case_contact.supervisor).deliver_later
+    end
+  end
+
+  def update_volunteer_address(case_contact)
+    return unless case_contact.volunteer_address.present? && !case_contact.address_field_disabled?
+
+    address = case_contact.volunteer.address || case_contact.volunteer.build_address
+    address.update(content: case_contact.volunteer_address)
```

Comment:
I'm still worried about this update silently failing if address is not valid for some reason. Can we use `update_columns` to avoid the validation check, or mark the case contact with an error if this address doesn't save correctly?

Reply:
There are no current validations on the `content` column for address. A lot of tests have something like `123 str`. But, I see your point if they are added in the future.

At this point in the code, you wouldn't things blowing up or getting validation errors, since the case contacts have already been split (see `create_additional_case_contacts(@case_contact)` on line 73), but I can add a validation check when the form is submitted. It will always pass now, but if validations are added to the `Address` class later, it should then use those.